rasterio/env.py

```diff
@@ -1,21 +1,25 @@
 """Rasterio's GDAL/AWS environment"""
 
 from functools import wraps
-import itertools as it
 import logging
+import threading
 
 from rasterio._env import (
     GDALEnv, del_gdal_config, get_gdal_config, set_gdal_config)
+from rasterio.compat import string_types
 from rasterio.dtypes import check_dtype
 from rasterio.errors import EnvError
-from rasterio.compat import string_types
 from rasterio.transform import guard_transform
 from rasterio.vfs import parse_path, vsi_path
 
 
 # The currently active GDAL/AWS environment is a private attribute.
-_env = None
+class ThreadEnv(threading.local):
+    def __init__(self):
+        self._env = None  # Initialises in each thread
+        self._discovered_options = None
 
+local = ThreadEnv()
 
 # When the outermost 'rasterio.Env()' executes '__enter__' it probes the
```

Comment:
Could you move this long comment closer to `_discovered_options`,  maybe just as a docstring on `ThreadEnv()`?  It may provide those dealing with both `osgeo` and Rasterio with some important background info.